### PR TITLE
Use modern typing annotations

### PR DESCRIPTION
**Pull request by perlinm**
Prompted by #1

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -65,7 +65,7 @@ extend-ignore = [
   "I201",
   "I202",
 ] # these errors conflict with black and/or isort
-type_checking_exempt_modules = ["collections.abc"]
+# type_checking_exempt_modules = ["collections.abc"] # REPO-SPECIFIC CONFIG
```
Comment (richrines1) [inline pyproject.toml:68]: why this change? putting this here just means that imports from `collections.abc` don't have to be in a TYPE_CHECKING block (as is already true for imports from `typing`), which means we don't need to put quotes around annotations as often

File: qldpc/objects.py [hunk 2]
```diff
@@ -17,12 +17,16 @@
...
-from typing import Collection, Literal, Optional
+from typing import TYPE_CHECKING
 
 import networkx as nx
 
 from qldpc import abstract
 
+if TYPE_CHECKING:
+    from collections.abc import Collection
+    from typing import Literal
```
Comment (richrines1) [inline qldpc/objects.py:28]: assuming the pyproject.toml change is reverted these shouldn't need to be in a type-checking block, and if then the quotes wouldn't be necessary around the annotations below

File: qldpc/objects.py [hunk 3]
```diff
@@ -227,9 +231,9 @@ def __init__(
     def get_min_rank(
         cls,
         group: abstract.Group,
-        subset_a: Collection[abstract.GroupMember],
-        subset_b: Collection[abstract.GroupMember],
-    ) -> Literal[0, 1, 2]:
+        subset_a: "Collection[abstract.GroupMember]",
+        subset_b: "Collection[abstract.GroupMember]",
+    ) -> "Literal[0, 1, 2]":
```
Comment (richrines1) [inline qldpc/objects.py:236]: (assuming prior)
```suggestion
        subset_a: Collection[abstract.GroupMember],
        subset_b: Collection[abstract.GroupMember],
    ) -> Literal[0, 1, 2]:
```
(ditto below)

**Review by richrines1 — COMMENTED** [review]
nice! a few recommendations:
1. add `from __future__ import annotations` to the top of every file to make sure this will work with older python versions
2. revert the change to `pyproject.toml` so imports from `collections.abc` don't have to be in a TYPE_CHECKING block
3. rerun `checks/flake8_.py`, which (assuming flake8-modern-annotations is installed) should flag the now-unnecessary quotes around annotations

**Comment by perlinm** [comment]
Awesome thanks @richrines1!  Recommendations implemented.

File: pyproject.toml (hunk 1 above)
Comment (perlinm) [reply pyproject.toml]: I guess I thought this exemption was added to avoid adding busywork (from converting annotation formats) to `*-superstaq`.

If you think it's good practice to import from `collections.abc` directly, that's good enough reason for me to leave this exemption in place.  My syntax highlighter is certain happier this way :smile:

File: qldpc/abstract.py [hunk 4]
```diff
@@ -239,7 +240,7 @@ def __eq__(self, other: object) -> bool:
     def __iter__(self) -> Iterator[tuple[GroupMember, "ModularInteger"]]:
         yield from self._vec.items()
 
-    def __add__(self, other: Union[GroupMember, "Element"]) -> "Element":
+    def __add__(self, other: "GroupMember | Element") -> "Element":
```
[the remark this answers is not in the document]
Comment (perlinm) [reply qldpc/abstract.py]: Ah nice.  I've always been a bit annoyed by the type checker not recognizing a class name within that class.

File: pyproject.toml (hunk 1 above)
Comment (richrines1) [reply pyproject.toml]: the main point of the type_checking block is to avoid module-level imports which are used for type checking but aren't necessary at runtime (which can be helpful for avoiding cyclic imports, or deferring costly imports until they're necessary (though we tend to avoid lazy imports so this application isn't particularly relevant to us))

it doesn't really make sense to e.g. put imports from `typing` in a type_checking block, because we'll have to load the `typing` module anyway just to import `TYPE_CHECKING` itself (and for this reason `flake8-type-checking` doesn't require imports from typing to be in a type_checking block). Similarly `collections.abc` is imported within `typing` anyway, so imo it doesn't make sense to put it in a type_checking block either (especially given its ubiquity and how much clumsier it makes the import syntax)

...so tldr imo `flake8-type-checking` is wrong to not exempt collections.abc by default

**Review by richrines1 — APPROVED** [review]
lgtm!

File: pyproject.toml (hunk 1 above)
Comment (perlinm) [reply pyproject.toml]: Cool makes total sense
